### ISSUE TITLE
Add method to filter for operational structures

### ISSUE THREAD

**Issue by DanRStevens** [OutpostUniverse/OPHD#755]
As suggested in: https://github.com/OutpostUniverse/OPHD/pull/753#discussion_r569296860

There should be a way to filter a list of structures for just the operational ones. That would eliminate the frequent need to check the operational status of structures inside of loops.

----

Note that `StructureList` is defined as:
```cpp
using StructureList = std::vector<Structure*>;
```

As the elements are just pointers, it should be fairly cheap to copy (reasonably sized) lists. That means copied lists could be filtered to remove non-operational structures, without all that much overhead.

Related: #754

----

If memory allocation efficiency is a concern, a wrapper list type could be created, where the iterator automatically skips over non-operational structures. This could be done embedded in the increment operator of the iterator type. That would mean the original data is still being traversed, so no need for copies or extra memory allocations.

Usage would look pretty similar to normal loops:
```cpp
const auto operationalStructures = OperationalStructures{structureManager.structureList(structureType)};

for (const auto& structure : operationalStructures) {
  // ...
}
```

**Comment by ldicker83** [comment]
I'm not sure memory allocation is much of an issue. Once the list goes out of scope the memory is released anyway and I'm not convinced memory fragmentation is going to be an issue for OPHD. If it ever does become a problem we can address it there.

I tend to prefer using references to lists but since these aren't calls being made every frame, just during turn processing, I can live with overhead involved in this.

My thinking would be the addition of a new function (either overloaded or specifically named), something like this:

```cpp
StructureList structureList(Structure::StructureClass, StructureState);

// or, perhaps preferred

StructureList structuresInState(Structure::StructureClass, StructureState);
```

Heavily leaning toward the second option. Should be a straight-forward change. Will slate it for the 0.8.2 milestone along with some code cleanup tasks using it.

**Comment by DanRStevens** [comment]
I suppose there would be a way to allow for a generic wrapper type for filtered lists. Could allow for a lambda to choose which elements to select or filter out.

You know, I think I might be starting to lean towards the low memory overhead option.

**Comment by DanRStevens** [comment]
If we allow for C++20, this could be done using [`std::ranges::filter_view`](https://en.cppreference.com/w/cpp/ranges/filter_view).

Would be used something like:
```cpp
const auto operationalStructures = structures | std::views::filter(Structure::operational)
```

Could wrap that into a method to make using it easier.

**Comment by ldicker83** [comment]
I don't see why we shouldn't use C++20. I suspect VS2019 already supports it. The big question mark here is Xcode on macos -- Apple has an annoying habit of being _way_ behind the times.

**Comment by ldicker83** [comment]
I found a reason why we can't use c++20.... MacOS fails horribly (as per usual) with new standards. I know we can have the compiler set to use the 'draft preview' or whatever they called it but as always Apple makes it hard to use bleeding edge stuff.

**Comment by Brett208** [comment]
I would agree you might want to wait another year or so before switching to C++20 to ease compatibility if possible to make sure all compilers are fully supporting it.

**Comment by DanRStevens** [comment]
Yeah, I did some test builds for NAS2D and OPHD. There were multiple failures:
https://app.circleci.com/pipelines/github/OutpostUniverse/OPHD?branch=updateCppStd
https://app.circleci.com/pipelines/github/lairworks/nas2d-core?branch=updateCppStd

Using `c++2a` solves some of these, though it looks like we'd need to drop the Ubuntu 18.04 LTS support.

I think I'm inclined to wait about a year to let compiler implementations catch up. Kind of a shame though.

Meanwhile we can either shelve the implementation idea for now, or write custom filter code (which would be compatible with c++17), and later dump it when we upgrade to c++20.

**Comment by belgianguy** [comment]
Since the project now runs in C++20, the implementation can be done, I tried using the filter and filter_view methods, but kept getting an error "error: invalid use of non-static member function ‘bool Structure::operational() const"

So I took to the forum and got some help from Arklon, who suggested to use a lambda instead. I made an effort to make it look like "allStructures".

`StructureList StructureManager::allOperationalStructures() 
{
	StructureList structuresOperational;
	auto operational = [](const Structure* x) { return x->operational(); };

	for (auto& classListPair : mStructureLists)
	{
		auto& structures = classListPair.second;
		std::copy_if(structures.begin(), structures.end(), std::back_inserter(structuresOperational), operational);
	}

	return structuresOperational;
}`

Would that suffice or is there another way to employ the filter and filter_view methods here?

**Comment by DanRStevens** [comment]
Ahh yes, I was thinking with C++20 features enabled, work on this could proceed soon.

So the code you wrote there would work, (and it would work with C++17), with the one downside of having dynamic memory allocation. The original collection of structures must be copied to a new collection, requiring dynamic memory allocation, then that collection is then returned, probably used in a range-for loop, and likely discarded soon after with the dynamic memory being freed. It's correct, but not necessarily efficient.

Consider the type of code this might be used to update:
```cpp
	for (auto structure : structureManager.structureList(structureClass))
	{
		if (structure->operational() || structure->isIdle())
		{
			storageCapacity += capacity;
		}
	}
```

If we examine that code, we see there are no dynamic memory allocations. The collection is returned by const reference and the `if` handles the filtering, without copying. However, this has quite the impact on the calling code. It would be nice to be able to handle the filtering as part of the API, without incurring extra dynamic memory allocation.

Consider this possible updated usage, which is perhaps easier to write and understand (especially with pre-prepared functions for common filtering operations):
```cpp
	const auto filterExpression = [](Structure& structure) { return structure.operational() || structure.isIdle(); };
	for (auto structure : structureManager.structureList(structureClass, filterExpression))
	{
		storageCapacity += capacity;
	}
```

Now, this could be made to work with C++17, but you'll likely still be returning a copied collection, and doing the filtering internally to the API call.

The benefit of C++20 and the [`ranges`](https://en.cppreference.com/w/cpp/header/ranges) library is you can now accomplish this without copying data to a new collection and incurring the cost of the dynamic memory allocation. One caveat to this, is that you're no longer returning a collection object, but rather a `view` object. The `view` object would contain reference to the underlying collection, and any filtering lambda, and would call it as needed to fetch the next item in the range-for loop. Effectively the structures in the collection are generated lazily on demand.

Using the `view` return type approach, we can improve and simplify the calling code, with no loss of efficiency.

----

The above is just a possible syntax. That might not be the way we want to go. Maybe we want to use separate wrapper objects, rather than putting the filter on the method call. Then we could filter structures from any source, not just that one method.
```cpp
	const auto filterExpression = [](Structure& structure) { return structure.operational() || structure.isIdle(); };
	for (auto structure : FilterStructures(structureManager.structureList(structureClass), filterExpression))
	{
		storageCapacity += capacity;
	}

```

----

In terms of how that's implemented, I'm not yet sure. I haven't looked at the `ranges` library code in quite a while. I suspect you'll need some kind of templated `view` object to be passed to the range-for. That would allow the specifics of the filter function to be compiled into the loop code, without any extra method calls. Though that does mean template instantiation, would could make for slower code.

Of course, if you fixed the filters, you could have specific pre-generated objects, and not need template instantiation.
```cpp
for (const auto& structure : OperationalFilter(structureList)) { /* ... */ }
```

It's also likely you could easily compose sub objects if they took a `view` as input to work on, which I assume a `std::vector` could be easily or implicitly converted to. Maybe you want operational factories that are producing something:
```cpp
for (const auto& structure : ProducingFilter(OperationalFilter(factoryList))) { /* ... */ }
```

**Comment by DanRStevens** [comment]
Without updating the `StructureManager` code, the example could be transformed into:
```cpp
const auto filterExpression = [](Structure* structure) { return structure->operational() || structure->isIdle(); };
for (auto structure : structureManager.structureList(structureClass) | std::views::filter(filterExpression))
{
	storageCapacity += capacity;
}
```

Maybe we should use the pipe syntax (`|`) to filter collections, rather than update the API to allow for filter parameters, or filtered versions of functions. I'm not really sure yet.
